src/librustc/traits/select.rs

```diff
@@ -2612,6 +2617,37 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                     &[inner_target]));
             }
 
+            // (.., T) -> (.., U).
+            (&ty::TyTuple(tys_a, _), &ty::TyTuple(tys_b, _)) => {
```

Comment:
I've thought before that we should have an "infinite family" of tuple ADTs. One complication is that ADTs have def-ids...